net/ipfilter/ipfilter.c

```diff
@@ -0,0 +1,632 @@
+
+      if (!ipfilter_match_proto(&filter->common, l4hdr, ipv4->proto))
+        {
+          continue;
+        }
+
+      /* Return the target action if matched. */
+
+      return filter->common.target;
+    }
+
+  /* Normally there should be a default rule in chain, won't reach here. */
+
+  ninfo("No filter matched, maybe uninitialized.\n");
+  return IPFILTER_TARGET_ACCEPT;
+}
+#endif
+
+#ifdef CONFIG_NET_IPv6
+static int ipv6_filter_match(FAR const struct net_driver_s *indev,
+                             FAR const struct net_driver_s *outdev,
+                             FAR const struct ipv6_hdr_s *ipv6,
+                             enum ipfilter_chain_e chain)
+{
+  FAR const struct ipv6_filter_entry_s *filter;
+  FAR const sq_queue_t *queue = &g_ipv6_filters[chain];
+  FAR const sq_entry_t *entry;
+  FAR const void *l4hdr;
+  uint8_t proto;
+  bool matched;
+
+  /* Handle unexpected status, return ACCEPT to indicate doing nothing. */
+
+  if ((indev == NULL && outdev == NULL) || ipv6 == NULL)
+    {
+      return IPFILTER_TARGET_ACCEPT;
+    }
+
+  l4hdr = IPv6_L4HDR(ipv6, proto);
+
+  sq_for_every(queue, entry)
+    {
+      filter = (FAR struct ipv6_filter_entry_s *)entry;
+
+      /* Match device */
+
+      if (!ipfilter_match_device(&filter->common, indev, outdev))
+        {
+          continue;
+        }
+
+      /* Match addresses */
+
+      matched = net_ipv6addr_maskcmp(filter->sip, ipv6->srcipaddr,
+                                     filter->smsk)
+                ^ filter->common.inv_srcip;
+      if (!matched)
+        {
+          continue;
+        }
+
+      matched = net_ipv6addr_maskcmp(filter->dip, ipv6->destipaddr,
+                                     filter->dmsk)
+                ^ filter->common.inv_dstip;
+      if (!matched)
+        {
+          continue;
+        }
+
+      /* Match protocol */
+
+      if (!ipfilter_match_proto(&filter->common, l4hdr, proto))
+        {
+          continue;
+        }
+
+      /* Return the target action if matched. */
+
+      return filter->common.target;
+    }
+
+  /* Normally there should be a default rule in chain, won't reach here. */
+
+  ninfo("No filter matched, maybe uninitialized.\n");
+  return IPFILTER_TARGET_ACCEPT;
+}
+#endif
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: ipfilter_cfg_alloc
+ *
+ * Description:
+ *   Allocate a new filter configuration entry for the given address family.
+ *
+ * Input Parameters:
+ *   family - The address family of the filter entry
+ *
+ * Returned Value:
+ *   A pointer to the newly allocated filter entry.  NULL is returned on
+ *   failure.
+ *
+ ****************************************************************************/
+
+FAR struct ipfilter_entry_s *ipfilter_cfg_alloc(sa_family_t family)
```

Comment:
remove this method if nobody using

Reply:
Thanks! I designed this for use in other places but forgot it. Now `convert_ipv4entry` and `convert_ipv6entry` will use it and we may optimize the allocation logic in this function later.